Add class for storing weight of cells

TheMarex opened the pull request:
# Issue

This class in an integral part of the MLD approach because it stores the weights of the overlay cells and indexes them by node ids. 

## Tasklist
 - [x] add unit tests
 - [x] review
 - [x] adjust for comments

oxidase commented on include/util/cell_storage.hpp:
```diff
@@ -0,0 +1,295 @@
+    {
+      private:
+        using WeightValueT = EdgeWeight;
+        using WeightRefT = decltype(*WeightPtrT());
```
this is misleading, i would suggest to use explicit type traits `std::remove_pointer` and `std::add_lvalue_reference` 
But better to change `WeightPtrT` to `WeightType` and than make  pointer or reference types out of `WeightType`

oxidase commented on include/util/cell_storage.hpp:
```diff
@@ -0,0 +1,295 @@
+
+        std::size_t GetRow(NodeID node) const
+        {
+            return std::find(source_boundary, source_boundary + num_source_nodes, node) -
```
shall it be asserted that `node` is in `source_boundary`?

oxidase commented on include/util/cell_storage.hpp:
```diff
@@ -0,0 +1,295 @@
+        {
+          public:
+            explicit ColumnIterator(WeightPtrT begin, std::size_t row_length)
+                : current(begin), offset(row_length)
```
`offset` -> `stride`. i would suggest to make later a cache-friendly storage

oxidase commented on include/util/cell_storage.hpp:
```diff
@@ -0,0 +1,295 @@
+                bool is_boundary_node = false;
+
+                for (auto edge = base_graph.BeginEdges(node); edge < base_graph.EndEdges(node);
+                     ++edge)
```
just minor: `for (auto edge : base_graph.GetAdjacentEdgeRange(node))`

daniel-j-h commented on include/util/cell_storage.hpp:
```diff
@@ -0,0 +1,295 @@
+
+    // Implementation of the cell view. We need a template parameter here
+    // because we need to derive a read-only and read-write view from this.
+    template <typename WeightPtrT> class CellImpl
```
Below: assert for `ptr` in ctor / functions, or take a ref

daniel-j-h replied:
Also if you only want this to work with pointers and not e.g. random access iterators

    static_assert(is_pointer<T>:value, "");

daniel-j-h commented on include/util/cell_storage.hpp:
```diff
@@ -0,0 +1,295 @@
+        const NodeID *const destination_boundary;
+
+        using RowIterator = WeightPtrT;
+        class ColumnIterator : public std::iterator<std::random_access_iterator_tag, EdgeWeight>
```
I would just use boost's iterator facade to implement this random access iterator; makes away with a lot boilerplate and actually provides things like required typedefs. Check the osrm code base we are already using it. And see

http://www.boost.org/doc/libs/1_55_0/libs/iterator/doc/iterator_facade.html

daniel-j-h replied:
On second thought, shouldn't a column iterator be the same as a strided iterator over your flattened matrix?

http://www.boost.org/doc/libs/1_55_0/libs/range/doc/html/range/reference/adaptors/reference/strided.html

TheMarex replied:
Added a comment to investigate this. I need to read the source code to verify whether this will have similar performance.

daniel-j-h commented on include/util/cell_storage.hpp:
```diff
@@ -0,0 +1,295 @@
+        }
+
+      public:
+        std::pair<RowIterator, RowIterator> GetOutWeight(NodeID node) const
```
Here and below: for all the pairs of iterators, what about either using

    boost:make_iterator_range(first, last)

or creating a small struct with `begin` and `end` functions so we can iterate over the return values and use std algorithms directly.

daniel-j-h commented on include/util/cell_storage.hpp:
```diff
@@ -0,0 +1,295 @@
+        CellImpl(const CellData &data,
+                 WeightPtrT const all_weight,
+                 const NodeID *const all_sources,
+                 const NodeID *const all_destinations)
```
Hm if you really want to user pointers at least assert them not being nullptrs. The interface is a bit ugly..

daniel-j-h commented on include/util/cell_storage.hpp:
```diff
@@ -0,0 +1,295 @@
+        }
+    };
+
+    inline std::size_t LevelIDToIndex(LevelID level) const { return level-1; }
```
These are all inline by default, this won't help at all

daniel-j-h commented on include/util/cell_storage.hpp:
```diff
@@ -0,0 +1,295 @@
+    using ConstCell = CellImpl<const EdgeWeight *>;
+
+    template <typename GraphT>
+    CellStorage(const MultiLevelPartition &partition, const GraphT &base_graph)
```
Realistically speaking, we want to use only the static graph here (?)

TheMarex replied:
DynamicGraph would work as well, considering we might run this after the graph transformation in the partitioner without creating a new graph. Maybe we should add a graph-trait to solidify the interface?

daniel-j-h replied:
The partitioner branch already has some graph traits
https://github.com/Project-OSRM/osrm-backend/blob/5ce90558ce535f02ffdcd589ffc515d0010d4220/include/util/graph_traits.hpp
mainly to allow to to SFINAE out edge `.data` functions and providing a `void` specialization.

You can add your own traits for all the functions we require in customization and then static assert for them, for sure.

daniel-j-h commented on include/util/cell_storage.hpp:
```diff
@@ -0,0 +1,295 @@
+            auto level_offset = level_to_cell_offset[LevelIDToIndex(level)];
+
+            std::vector<std::pair<CellID, NodeID>> level_source_boundary;
+            std::vector<std::pair<CellID, NodeID>> level_destination_boundary;
```
Hm this creates vectors and destroys them after each iteration again. Should be fine with only a couple of levels, but still you could hoist them outside and `.clear()` between each iteration instead.

daniel-j-h commented on include/util/cell_storage.hpp:
```diff
@@ -0,0 +1,295 @@
+            }
+
+            std::sort(level_source_boundary.begin(), level_source_boundary.end());
+            std::sort(level_destination_boundary.begin(), level_destination_boundary.end());
```
tbb sort?

daniel-j-h commented on include/util/for_each_range.hpp:
```diff
@@ -0,0 +1,27 @@
+    {
+        const auto key = iter->first;
+        auto begin_range = iter;
+        while (iter != end && iter->first == key)
```
Not sure if this should be a generic util function then at all. Even now it depends on the sub ranges being represented in pairs.

daniel-j-h replied:
Also isn't this simply

    for_each(f, l, [](auto sub_f, auto sub_l) { for_each(sub_f, sub_f, functor); };

at the moment

daniel-j-h commented on include/util/cell_storage.hpp:
```diff
@@ -0,0 +1,295 @@
+            weight_offset += cell.num_source_nodes * cell.num_destination_nodes;
+        }
+
+        weights.resize(weight_offset+1, INVALID_EDGE_WEIGHT);
```
clang-format

daniel-j-h commented on include/util/cell_storage.hpp:
```diff
@@ -0,0 +1,295 @@
+    {
+        auto offset = level_to_cell_offset[LevelIDToIndex(level)];
+        return Cell{cells[offset + id], &weights[0], &source_boundary[0], &destination_boundary[0]};
+    }
```
Would love some assertions in the file - especially since you subscript tons of vectors.

daniel-j-h commented on include/util/multi_level_partition.hpp:
```diff
@@ -0,0 +1,33 @@
+using CellID = std::uint32_t;
+
+// Mock interface, can be removed when we have an actual implementation
+class MultiLevelPartition
```
What is the idea behind making this a abstract base class?

Can't this be a concept/trait like interface with concrete implementations?
We don't need runtime polymorphism here, do we?

TheMarex replied:
No runtime polymorphism will be removed, see comment. This is just during implementation. I didn't want to couple building this class and the actual MultiLevelPartition.

TheMarex commented on include/util/cell_storage.hpp:
```diff
@@ -0,0 +1,295 @@
+                    const auto &data = base_graph.GetEdgeData(edge);
+
+                    is_boundary_node |= partition.GetCell(level, other) != cell_id;
+                    is_source_node |= partition.GetCell(level, other) == cell_id && data.forward;
```
What we are checking here is whether this node will have a row /column in the cell that could contain anything other than `INVALID_EDGE_WEIGHT`. This is only the case if there is any path starting from 0 in the cell, that could eventually lead to another border node.

There is no path in the blue cell, leading from 1 to 0. So if you would add 0 as a destination that would be a column that all weights are set to `INVALID_EDGE_WEIGHT`. However we there is an edge going from 0 to 1 so there could be a path to another border vertex (also `1` in this case).

TheMarex commented:
@oxidase @daniel-j-h can you take a look again if I addressed your concerns appropriately? Currently fighting a little bit of travis weirdness but it should be good to go after that.

daniel-j-h commented on include/util/cell_storage.hpp:
```diff
@@ -0,0 +1,314 @@
+            explicit ColumnIterator(WeightPtrT begin, std::size_t row_length)
+                : current(begin), stride(row_length)
+            {
+            }
```
assert(begin)

daniel-j-h commented on include/util/cell_storage.hpp:
```diff
@@ -0,0 +1,314 @@
+
+            bool operator!=(const ColumnIterator &other) const { return current != other.current; }
+
+            int operator-(const ColumnIterator &other) const
```
Why int as a return type here?

TheMarex replied:
Right this should probably be a signed 64bit integer. 👍

daniel-j-h commented on include/util/cell_storage.hpp:
```diff
@@ -0,0 +1,314 @@
+        cells.resize(number_of_cells);
+
+        std::vector<std::pair<CellID, NodeID>> level_source_boundary;
+        std::vector<std::pair<CellID, NodeID>> level_destination_boundary;
```
What about making this proper types instead of using `.first` and `.second` below

    struct LevelNode { CellID level; NodeID node; };

    std::vector<LevelNode> sources, sinks;

Then we can say `.level` and `.node` instead.

TheMarex replied:
For temporary types like this, I prefer using pair/tuple since they automatically derive all of the operators needed. I hope variable names such as `cell_and_node` compensate for that.

oxidase replied:
why not in local name scope `using LevelNode = std::pair<CellID, NodeID>;`?

daniel-j-h commented on include/util/cell_storage.hpp:
```diff
@@ -0,0 +1,314 @@
+                    if (is_source_node)
+                        level_source_boundary.emplace_back(cell_id, node);
+                    if (is_destination_node)
+                        level_destination_boundary.emplace_back(cell_id, node);
```
Hm can we guestimate the number of sources and sinks here and `.reserve` above?

TheMarex replied:
It is hard to estimate and depends heavily on the quality of the partition. I think 5% of all nodes is a good estimate for level 1, and every level above that will be much below this anyway.

daniel-j-h commented on include/util/cell_storage.hpp:
```diff
@@ -0,0 +1,314 @@
+                                   end,
+                                   std::back_inserter(destination_boundary),
+                                   [](const auto &cell_and_node) { return cell_and_node.second; });
+                });
```
This and above shared a large amount of code - what about abstracting over it?

daniel-j-h commented on include/util/cell_storage.hpp:
```diff
@@ -0,0 +1,314 @@
+        BOOST_ASSERT(cell_index < cells.size());
+        return Cell{
+            cells[cell_index], weights.data(), source_boundary.data(), destination_boundary.data()};
+    }
```
Hm const/non-const has the same impl. - what about const_cast'ing and implementing one wrt. the other?
Check Meyer's Effective C++ for how this can be done.

TheMarex replied:
The problem I tried to solve here is around read-only views and read-write views (we need both). Even if a `Cell` itself can be `const` it might return non-const iterators to the backing store. All of the methods of `Cell` are always `const`, the view cannot be changed. I don't see how we could use `const` overloading there tor return appropriate iterators in this case?
